lib/cloud-metadata/gcp.js

```diff
@@ -114,11 +108,11 @@ function formatMetadataStringIntoObject (string) {
 
   if (data.instance) {
     metadata.instance = {
-      id: data.instance.id + ''
+      id: String(data.instance.id)
     }
 
     metadata.machine = {
-      type: (data.instance.machineType.split('/').pop()) + ''
+      type: String((data.instance.machineType.split('/').pop()))
```

Comment:
All these stringifications via either `String()` or `+ ''` feel odd. Was there a particular case where a `null` or a number cause surprise? Would it be cleaner to handle just that one case rather than blanket String-ing everything? I don't have a particular case, but could `String(allTheThings)` possibly mask a bug in a value erroneously being `null` or `undefined`?

Reply:
APM Server requires these values to be strings.  If we send a non-string where a string is expected, APM Server will reject the metadata payload and no data will be sent.  APM Server's schema is here: https://github.com/elastic/apm-server/blob/d8c06ea35822476838bcb4d7cde4db50c79798c1/docs/spec/v2/metadata.json

I discovered this with the GCP instance data API. There were a few cases where we got Numbers instead of strings. That's what started the casting. 

This made me realize that there's no explicit contract about what types will be returned from these instance data APIs. Just because they return strings now does not mean they'll return strings in the future. The only way to guarantee sending APM Server the strings it wants was to explicitly cast somewhere.   

Casting a null or undefined _might_ introduce a bug where strange data was sent through. However -- the trade off of not casting is that APM Server rejects the agent's data and users don't see anything.  Also -- the alternatives for a null or undefined would be to send either an empty string, or a not send an expected field at all, which also introduces the possibility of a bug.   

Open to other solutions/suggestions here -- but the limiting factor seems to be that APM Server expects certain datatypes and we can't guarantee we'll get those types.  Also, just to say it out loud, perhaps _the right_ solution here would be to collect whatever the instance metadata servers send us, but then cast and verify the data based on the schemas we have for APM Server immediately before encoding/sending it.  That is a possibility -- but seems like something we might want to tackle at a later point (time being the ultimate tradeoff)

Reply:
Thanks. +1 to what you have.